hypothesis-python/tests/array_api/test_arrays.py

```diff
@@ -10,145 +10,148 @@
 
 import pytest
 
-from hypothesis import assume, given, strategies as st
+from hypothesis import given, strategies as st
 from hypothesis.errors import InvalidArgument
-from hypothesis.extra.array_api import DTYPE_NAMES, NUMERIC_NAMES
+from hypothesis.extra.array_api import DTYPE_NAMES, NUMERIC_NAMES, UINT_NAMES
 from hypothesis.internal.floats import width_smallest_normals
 
-from tests.array_api.common import COMPLIANT_XP, WIDTHS_FTZ, xp, xps
-from tests.common.debug import find_any, minimal
-from tests.common.utils import fails_with, flaky
-
-needs_xp_unique_values = pytest.mark.skipif(
-    not hasattr(xp, "unique_values"), reason="optional API"
-)
+from tests.array_api.common import flushes_to_zero, skip_on_unsupported_dtype
+from tests.common.debug import assert_all_examples, find_any, minimal
+from tests.common.utils import flaky
 
 
-# xp.unique_value() should return distinct NaNs - if not, tests that (rightly)
-# assume such behaviour will likely fail. This mark namely addresses mocking the
-# array module with NumPy 1.21, which treats NaNs as not distinct.
-# See https://mail.python.org/pipermail/numpy-discussion/2021-August/081995.html
-two_nans = xp.asarray([float("nan"), float("nan")])
-assumes_distinct_nans = pytest.mark.xfail(
-    hasattr(xp, "unique_values") and xp.unique_values(two_nans).size != 2,
-    reason="NaNs not distinct",
-)
+def skip_on_missing_unique_values(xp):
+    if not hasattr(xp, "unique_values"):
+        pytest.mark.skip("optional API")
 
 
-def assert_array_namespace(x):
-    """Check array has __array_namespace__() and it returns the correct module.
-
-    This check is skipped if a mock array module is being used.
+def xfail_on_indistinct_nans(xp):
+    """
+    xp.unique_value() should return distinct NaNs - if not, tests that (rightly)
+    assume such behaviour will likely fail. This mark namely addresses mocking
+    the array module with NumPy 1.21, which treats NaNs as not distinct.
+    See https://mail.python.org/pipermail/numpy-discussion/2021-August/081995.html
     """
-    if COMPLIANT_XP:
-        assert x.__array_namespace__() is xp
+    skip_on_missing_unique_values(xp)
+    two_nans = xp.asarray([float("nan"), float("nan")])
+    if xp.unique_values(two_nans).size != 2:
+        pytest.xfail("NaNs not distinct")
 
 
-@given(xps.scalar_dtypes(), st.data())
-def test_draw_arrays_from_dtype(dtype, data):
+@pytest.mark.parametrize("dtype_name", DTYPE_NAMES)
+def test_draw_arrays_from_dtype(xp, xps, dtype_name):
     """Draw arrays from dtypes."""
-    x = data.draw(xps.arrays(dtype, ()))
-    assert x.dtype == dtype
-    assert_array_namespace(x)
+    try:
+        dtype = getattr(xp, dtype_name)
+    except AttributeError:
+        pytest.skip(f"dtype {dtype_name} not supported")
```

Comment:
I don't think that we should skip on missing dtypes if they're required by the standard, at least not by default.

(happy to add an environment variable of array-apis-to-allow-skipping-dtypes, which you could set to e.g. `...=pytorch_wrapper` to allow for known issues in some implementations without weakening the tests for others)